### DISCUSSION TITLE
Confusion on using wait() vs. add_callback()

### DISCUSSION THREAD

**Discussion by MitchCayCRC** [JoelBender/bacpypes#408] [Q&A] [answered]
Hi everyone,

I am still relatively new to bacpypes and am generally confused on knowing when to implement wait() vs add_callback() and also feel I may not be implementing them correctly? 

I have been building most of my applications with add_callback() and RecurringTask, everything was working fine until I started trying to push the task to be faster or add a set_timeout(). Initially with the set_timeout() issue my ioResponses were becoming out of sync after a timeoutError, I looked into #333 and thought maybe I should not be using add_callback and attempted to switch to wait()... However, whenever I use wait() my application seems to get stuck waiting for request_io() to complete. 

I recently switched back to using add_callback() and find it much more useful (even if I cant use a timeout when setting a callback) but don't like the fact that the fact that the application will continue on to the next task before the callback function completes.  Essentially is there a way to guarantee:
```
        # make an IOCB
        iocb = IOCB(request)
        
        # set a callback for the response
        iocb.add_callback(self.read_complete)
        
        # give it to the application
        self.request_io(iocb)
        
        print("on to next task")
             
    def read_complete(self, iocb):
        
        if iocb.ioResponse:
            print("completing callback function")
```
will always print "completing callback function" before "on to next task"? (also help with add_timeout would be appreciated!)

**Comment by JoelBender** [comment, marked as answer]
The callback function is only called when the request has been _completely_ processed by the controller, before that it might be sitting in a queue or only be partially processed.  When you give an iocb to a controller it is no longer yours, like a buffer that you have given to the old school `fgets()` and `fputs()` stdio functions.  You can look at the status code (polling-like) and do something with the results, pass a callback like you are doing now, or in a multi-threaded application _- - not recommended - -_ call `wait()` and your thread will be suspended until the event (semaphore) is set (signaled).

If an IOController has been given an IOCB instance, then in theory it is also responsible for doing whatever needs to happen when the request has been aborted, but given the nature of the client and server transaction state machines that manage the state of a request/response "transaction" that's been hard to get right.  Using the `set_timeout()` function is scheduling a call (an instance of a Task) to abort the IOCB request at some time in the future, whenever the IO loop in `run()` can get some time to run it.

I know exactly what you're looking for: `read_property_ack = await self.request(read_property_request)` but that will have to wait for **bacpypes3**.

In the mean time, check out the **DiscoverToDo.py** sample application with the `ToDoList` and `ToDoItem` classes.  Each ToDoItem instance gets a call to `prepare()` for some activity and return an IOCB, and then will be notified when the IO request is `complete()`.  For the discovery process, this means reading the object list and making a bunch of other ToDoItem instances for each object identifier in the result. When the item is complete, the next item is picked up out of the list and given its chance to `prepare()` and return a new IOCB.

**Reply by MitchCayCRC** [reply]
Thanks Joel this is extremely helpful! 

Since we're waiting on **bacpypes3** looking at **DiscoverToDo.py** gives me a very good idea of how I could approach this problem. In trying to better understand it though I am attempting to track the processing of each individual request to better understand how everything fits together. I am using the `--debug` flag to do so but scrolling through the console output is not the most efficient. Is there a simple way to print the debug logging out to a file for inspection?

**Reply by JoelBender** [reply]
If you say `--debug __main__:snork.txt` the output will go to the text file, with `--debug __main__:snork.txt:10485760` will output to rolling log files 10MB each.  You can also `export BACPYPES_DEBUG_FILE=snork.txt` in your shell, which is easier than added a file name to each of the pieces, unless you want to separate them for some other reason.
